In [1]:
import numpy as np
import pandas as pd
import os

np.random.seed(10)
os.getcwd()

'C:\\Users\\adity\\Downloads\\ML'

In [84]:
data = pd.read_csv('ENB2012_data.csv', na_values=['?'])
data.head()

X1     X2     X3      X4   X5  X6   X7  X8     Y1     Y2
0  0.98  514.5  294.0  110.25  7.0   2  0.0   0  15.55  21.33
1  0.98  514.5  294.0  110.25  7.0   3  0.0   0  15.55  21.33
2  0.98  514.5  294.0  110.25  7.0   4  0.0   0  15.55  21.33
3  0.98  514.5  294.0  110.25  7.0   5  0.0   0  15.55  21.33
4  0.90  563.5  318.5  122.50  7.0   2  0.0   0  20.84  28.28

In [3]:
data.describe()

X1          X2          X3          X4         X5          X6  \
count  768.000000  768.000000  768.000000  768.000000  768.00000  768.000000   
mean     0.764167  671.708333  318.500000  176.604167    5.25000    3.500000   
std      0.105777   88.086116   43.626481   45.165950    1.75114    1.118763   
min      0.620000  514.500000  245.000000  110.250000    3.50000    2.000000   
25%      0.682500  606.375000  294.000000  140.875000    3.50000    2.750000   
50%      0.750000  673.750000  318.500000  183.750000    5.25000    3.500000   
75%      0.830000  741.125000  343.000000  220.500000    7.00000    4.250000   
max      0.980000  808.500000  416.500000  220.500000    7.00000    5.000000   

               X7         X8          Y1          Y2  
count  768.000000  768.00000  768.000000  768.000000  
mean     0.234375    2.81250   22.307201   24.587760  
std      0.133221    1.55096   10.090196    9.513306  
min      0.000000    0.00000    6.010000   10.900000  
25%      0.100000    1.75000   12.992500   15.620000  
50%      0.250000    3.00000   18.950000   22.080000  
75%      0.400000    4.00000   31.667500   33.132500  
max      0.400000    5.00000   43.100000   48.030000

In [4]:
percentage_missing = (data.isnull().sum()*100)/(len(data))
print(percentage_missing)

X1    0.0
X2    0.0
X3    0.0
X4    0.0
X5    0.0
X6    0.0
X7    0.0
X8    0.0
Y1    0.0
Y2    0.0
dtype: float64


In [85]:
#Defining X and Y variables
x = data.drop(["Y1","Y2"], axis = 1)
y1 = data["Y1"]
y2 = data["Y2"]

## Predictng for Y1 - Heating Load

In [86]:
# Randomising data
from sklearn.model_selection import train_test_split
x_train, x_test, y1_train, y1_test = train_test_split(x, y1, test_size = 0.1, random_state = 10)

In [87]:
from sklearn.preprocessing import MinMaxScaler
mm_x = MinMaxScaler()
x_train = mm_x.fit_transform(x_train)
x_test = mm_x.fit_transform(x_test)

#### Importing Libraries for Score and GridSearch

In [8]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

##### Linear Regression

In [9]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
linReg_heat = regressor.fit(x_train, y1_train)

print('Linear Regression R-squared score (training) for Heating Load: {:.3f}'
     .format(linReg_heat.score(x_train, y1_train)))
print('Linear Regression R-squared score (test)  for Heating Load: {:.3f}'
     .format(linReg_heat.score(x_test, y1_test)))

Linear Regression R-squared score (training) for Heating Load: 0.914
Linear Regression R-squared score (test)  for Heating Load: 0.933


#### Support Vector Machine - Linear

In [10]:
from sklearn.svm import SVR
svr = SVR(kernel = 'linear')
param_grid = {'C':[0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100],
              'gamma' : [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100]}
grid_search3= GridSearchCV(svr,param_grid, cv=5)
grid_search3.fit(x_train, y1_train)

print("Best Parameters : {}".format(grid_search3.best_params_))
print("Best cross_val score : {}".format(grid_search3.best_score_))

Best Parameters : {'C': 100, 'gamma': 0.001}
Best cross_val score : 0.9077288977147542


In [11]:
svrlin = SVR(kernel = 'linear', C = 100, gamma = 0.001)
svrlin_heat = svrlin.fit(x_train,y1_train)
print('SVM Linear R-squared score (training) for Heating Load: {:.3f}'
     .format(svrlin_heat.score(x_train, y1_train)))
print('SVM Linear R-squared score (test) for Heating Load: {:.3f}'
     .format(svrlin_heat.score(x_test, y1_test)))

SVM Linear R-squared score (training) for Heating Load: 0.911
SVM Linear R-squared score (test) for Heating Load: 0.934


#### Support Vector Machine - Kernel

In [12]:
from sklearn.svm import SVR
svr2 = SVR(kernel = 'rbf')
param_grid = {'C':[0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100],
              'gamma' : [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100]}
grid_search4= GridSearchCV(svr2,param_grid, cv=5)
grid_search4.fit(x_train, y1_train)

print("Best Parameters : {}".format(grid_search4.best_params_))
print("Best cross_val score : {}".format(grid_search4.best_score_)) 

Best Parameters : {'C': 50, 'gamma': 10}
Best cross_val score : 0.9621492547474835


In [13]:
svrrbf= SVR(kernel = 'rbf', C = 50, gamma = 10)
svrrbf_heat = svrrbf.fit(x_train,y1_train)
print('R-squared score (training): {:.3f}'
     .format(svrrbf_heat.score(x_train, y1_train)))
print('R-_squared score (test): {:.3f}'
     .format(svrrbf_heat.score(x_test, y1_test)))

R-squared score (training): 1.000
R-_squared score (test): 0.972


#### Knn Regressor

In [14]:
from sklearn.neighbors import KNeighborsRegressor
knnreg = KNeighborsRegressor()
param_grid = {'n_neighbors': [3, 4, 5, 6, 7]}
grid_search_knnreg = GridSearchCV(knnreg, param_grid, cv = 5, scoring = 'r2')
grid_search_knnreg.fit(x_train, y1_train)
print("Best Parameters for KNN repressor - Heating Load: {}".format(grid_search_knnreg.best_params_))
print("Best Score for KNN repressor - Heating Load: {:.2f}".format(grid_search_knnreg.best_score_))

Best Parameters for KNN repressor - Heating Load: {'n_neighbors': 7}
Best Score for KNN repressor - Heating Load: 0.92


In [15]:
knn = KNeighborsRegressor(n_neighbors=7)
knnReg = knn.fit(x_train,y1_train)

print('R-squared score (training): {:.3f}'
     .format(knnReg.score(x_train, y1_train)))
print('R-squared score (test): {:.3f}'
     .format(knnReg.score(x_test, y1_test)))

R-squared score (training): 0.942
R-squared score (test): 0.932


#### After running OLS, SVR(Linear and Kernel), and KNN regressors, we can say that Linear SVR and OLS give very good results. Let us try bagging and boosting methods for these models

#### Adaboost Gridsearch

In [29]:
from sklearn.ensemble import AdaBoostRegressor

adaboost = AdaBoostRegressor(random_state = 10)
param_grid = {'n_estimators':[10,50,100,200,500,750,1000],
              'learning_rate':[0.1, 0.05, 0.01, 0.005, 0.001, 0.0005, 0.0001]}
grid_search_adaboost= GridSearchCV(adaboost,param_grid, cv=5, scoring = 'r2')
grid_search_adaboost.fit(x_train, y1_train)

print("Best Parameters : {}".format(grid_search_adaboost.best_params_))
print("Best cross_val score : {}".format(grid_search_adaboost.best_score_))

Best Parameters : {'learning_rate': 0.1, 'n_estimators': 500}
Best cross_val score : 0.9619858992725867


#### Adaboost for Linear Regression

In [30]:
ada_lin = AdaBoostRegressor(regressor, learning_rate = 0.1, n_estimators = 500, random_state = 10)
ada_lin.fit(x_train, y1_train)

print('Adaboost for Linear Regression on training set: {:.2f}'
     .format(ada_lin.score(x_train, y1_train)))
print('Adaboost for Linear Regression on test set: {:.2f}\n'
     .format(ada_lin.score(x_test, y1_test)))

Adaboost for Linear Regression on training set: 0.91
Adaboost for Linear Regression on test set: 0.93



#### Adaboost for Linear SVM

In [31]:
ada_svrlin = AdaBoostRegressor(svrlin, learning_rate = 0.1, n_estimators = 500, random_state = 10)
ada_svrlin.fit(x_train, y1_train)

print('Adaboost for Linear SVM on training set: {:.2f}'
     .format(ada_svrlin.score(x_train, y1_train)))
print('Adaboost for Linear SVM on test set: {:.2f}\n'
     .format(ada_svrlin.score(x_test, y1_test)))

Adaboost for Linear SVM on training set: 0.91
Adaboost for Linear SVM on test set: 0.93



#### Adaboost for Kernel SVM

In [32]:
ada_svrrbf = AdaBoostRegressor(svrrbf, learning_rate = 0.1, n_estimators = 500, random_state = 10)
ada_svrrbf.fit(x_train, y1_train)

print('Adaboost for Kernel SVM on training set: {:.2f}'
     .format(ada_svrrbf.score(x_train, y1_train)))
print('Adaboost for Kernel SVM on test set: {:.2f}\n'
     .format(ada_svrrbf.score(x_test, y1_test)))

Adaboost for Kernel SVM on training set: 1.00
Adaboost for Kernel SVM on test set: 0.96



#### Adaboost for KNN

In [33]:
ada_knn = AdaBoostRegressor(knn, learning_rate = 0.1, n_estimators = 500, random_state = 10)
ada_knn.fit(x_train, y1_train)

print('Adaboost for KNN on training set: {:.2f}'
     .format(ada_knn.score(x_train, y1_train)))
print('Adaboost for KNN on test set: {:.2f}\n'
     .format(ada_knn.score(x_test, y1_test)))


Adaboost for KNN on training set: 0.95
Adaboost for KNN on test set: 0.93



#### AdaBoost gives very good results for KNN compared to all other models. The train scores for OLS and Linear regression is less compared to KNN for the same test score.

#### Bagging Gridsearch

In [34]:
from sklearn.ensemble import BaggingRegressor

bagging = BaggingRegressor(random_state = 10)
param_grid = {'max_samples':[10,50,100]}
grid_search_bagging = GridSearchCV(bagging, param_grid, cv=5, scoring = 'r2')
grid_search_bagging.fit(x_train, y1_train)

print("Best Parameters : {}".format(grid_search_bagging.best_params_))
print("Best cross_val score : {}".format(grid_search_bagging.best_score_))

Best Parameters : {'max_samples': 100}
Best cross_val score : 0.9906340747620854


#### Bagging for Linear Regression

In [22]:
bag_lin= BaggingRegressor(regressor, max_samples = 100, random_state = 10)
bag_lin.fit(x_train, y1_train)

print('Bagging for Linear Regression on training set: {:.2f}'
     .format(bag_lin.score(x_train, y1_train)))
print('Bagging for Linear Regression on test set: {:.2f}\n'
     .format(bag_lin.score(x_test, y1_test)))

Bagging for Linear Regression on training set: 0.91
Bagging for Linear Regression on test set: 0.93



#### Bagging for Linear SVM

In [23]:
bag_svrlin= BaggingRegressor(svrlin, max_samples = 100, random_state = 10)
bag_svrlin.fit(x_train, y1_train)

print('Bagging for Linear SVM on training set: {:.2f}'
     .format(bag_svrlin.score(x_train, y1_train)))
print('Bagging for Linear SVM on test set: {:.2f}\n'
     .format(bag_svrlin.score(x_test, y1_test)))

Bagging for Linear SVM on training set: 0.91
Bagging for Linear SVM on test set: 0.93



#### Bagging for Kernel SVM

In [24]:
bag_svrrbf= BaggingRegressor(svrrbf, max_samples = 100, random_state = 10)
bag_svrrbf.fit(x_train, y1_train)

print('Bagging for Kernel SVM on training set: {:.2f}'
     .format(bag_svrrbf.score(x_train, y1_train)))
print('Bagging for Kernel SVM on test set: {:.2f}\n'
     .format(bag_svrrbf.score(x_test, y1_test)))

Bagging for Kernel SVM on training set: 0.77
Bagging for Kernel SVM on test set: 0.68



#### Bagging for KNN

In [25]:
bag_knn= BaggingRegressor(knn, max_samples = 100, random_state = 10)
bag_knn.fit(x_train, y1_train)

print('Bagging for Kernel SVM on training set: {:.2f}'
     .format(bag_knn.score(x_train, y1_train)))
print('Bagging for Kernel SVM on test set: {:.2f}\n'
     .format(bag_knn.score(x_test, y1_test)))

Bagging for Kernel SVM on training set: 0.90
Bagging for Kernel SVM on test set: 0.91



#### Looking at all the models built using ensemble methods, we can see that the scores of OLS and Linear SVR are consistent. We can use Linear SVR to predict the value of y1.

In [26]:
from sklearn.metrics import r2_score

y1_train_predict_svclin = bag_svrlin.predict(x_train)
y1_test_predict_svclin = bag_svrlin.predict(x_test)

print('Train r2: {:.2f}'.format(r2_score(y1_train, y1_train_predict_svclin)))
print('Test r2: {:.2f}'.format(r2_score(y1_test, y1_test_predict_svclin)))

Train r2: 0.91
Test r2: 0.93


#### Deep Learning Regressor

In [88]:
x_train.shape

(691, 8)

In [90]:
# create model
from keras.models import Sequential
from keras.layers import Dense
model = Sequential()
model.add(Dense(13, input_dim=8, kernel_initializer='normal', activation='relu'))
model.add(Dense(6, kernel_initializer = 'normal', activation = 'relu'))
model.add(Dense(1, kernel_initializer='normal'))

# Compile model
model.compile(loss='mse', optimizer='adam' , metrics = ['mse'])

model.fit(x_train, y1_train, epochs = 100, batch_size = 20)

Epoch 1/100
691/691 [==============================] - 3s 4ms/step - loss: 601.0948 - mean_squared_error: 601.0948
Epoch 2/100
691/691 [==============================] - 0s 118us/step - loss: 597.1087 - mean_squared_error: 597.1087
Epoch 3/100
691/691 [==============================] - 0s 122us/step - loss: 585.9691 - mean_squared_error: 585.9691
Epoch 4/100
691/691 [==============================] - 0s 139us/step - loss: 558.2864 - mean_squared_error: 558.2864
Epoch 5/100
691/691 [==============================] - 0s 131us/step - loss: 504.5510 - mean_squared_error: 504.5510
Epoch 6/100
691/691 [==============================] - 0s 135us/step - loss: 422.2890 - mean_squared_error: 422.2890
Epoch 7/100
691/691 [==============================] - 0s 129us/step - loss: 319.7315 - mean_squared_error: 319.7315
Epoch 8/100
691/691 [==============================] - 0s 133us/step - loss: 217.9224 - mean_squared_error: 217.9224
Epoch 9/100
691/691 [==============================] - 0s 363us/st

691/691 [==============================] - 0s 145us/step - loss: 9.8465 - mean_squared_error: 9.8465
Epoch 72/100
691/691 [==============================] - 0s 147us/step - loss: 9.7429 - mean_squared_error: 9.7429
Epoch 73/100
691/691 [==============================] - 0s 145us/step - loss: 9.7090 - mean_squared_error: 9.7090
Epoch 74/100
691/691 [==============================] - 0s 131us/step - loss: 9.6962 - mean_squared_error: 9.6962
Epoch 75/100
691/691 [==============================] - 0s 113us/step - loss: 9.6367 - mean_squared_error: 9.6367
Epoch 76/100
691/691 [==============================] - 0s 126us/step - loss: 9.6227 - mean_squared_error: 9.6227
Epoch 77/100
691/691 [==============================] - 0s 125us/step - loss: 9.5987 - mean_squared_error: 9.5987
Epoch 78/100
691/691 [==============================] - 0s 123us/step - loss: 9.5761 - mean_squared_error: 9.5761
Epoch 79/100
691/691 [==============================] - 0s 241us/step - loss: 9.5349 - mean_squared_e

In [91]:
from sklearn.metrics import r2_score

y1_train_predict = model.predict(x_train)
y1_test_predict = model.predict(x_test)

print('Train r2: {:.2f}'.format(r2_score(y1_train, y1_train_predict)))
print('Test r2: {:.2f}'.format(r2_score(y1_test, y1_test_predict)))

Train r2: 0.91
Test r2: 0.91


#### Compared to all the models, linear SVR is more robust and gives high test scores with comparable training scores. Hence, we can use Linear SVR to predict y1

## Predictng for Y2 - Cooling Load

In [35]:
# Randomising data
from sklearn.model_selection import train_test_split
x_train, x_test, y2_train, y2_test = train_test_split(x, y2, test_size = 0.1, random_state = 10)

In [36]:
from sklearn.preprocessing import MinMaxScaler
mm_x = MinMaxScaler()
x_train = mm_x.fit_transform(x_train)
x_test = mm_x.fit_transform(x_test)

#### Importing Libraries for Score and GridSearch

In [37]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

##### Linear Regression

In [38]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
linReg_heat = regressor.fit(x_train, y2_train)

print('Linear Regression R-squared score (training) for Heating Load: {:.3f}'
     .format(linReg_heat.score(x_train, y2_train)))
print('Linear Regression R-squared score (test)  for Heating Load: {:.3f}'
     .format(linReg_heat.score(x_test, y2_test)))

Linear Regression R-squared score (training) for Heating Load: 0.885
Linear Regression R-squared score (test)  for Heating Load: 0.909


#### Support Vector Machine - Linear

In [39]:
from sklearn.svm import SVR
svr = SVR(kernel = 'linear')
param_grid = {'C':[0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100],
              'gamma' : [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100]}
grid_search3= GridSearchCV(svr,param_grid, cv=5)
grid_search3.fit(x_train, y2_train)

print("Best Parameters : {}".format(grid_search3.best_params_))
print("Best cross_val score : {}".format(grid_search3.best_score_)) 

Best Parameters : {'C': 10, 'gamma': 0.001}
Best cross_val score : 0.8755295597031377


In [40]:
svrlin = SVR(kernel = 'linear', C = 10, gamma = 0.001)
svrlin_heat = svrlin.fit(x_train,y2_train)
print('SVM Linear R-squared score (training) for Heating Load: {:.3f}'
     .format(svrlin_heat.score(x_train, y2_train)))
print('SVM Linear R-squared score (test) for Heating Load: {:.3f}'
     .format(svrlin_heat.score(x_test, y2_test)))

SVM Linear R-squared score (training) for Heating Load: 0.877
SVM Linear R-squared score (test) for Heating Load: 0.908


#### Support Vector Machine - Kernel

In [40]:
from sklearn.svm import SVR
svr2 = SVR(kernel = 'rbf')
param_grid = {'C':[0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100],
              'gamma' : [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 10, 50, 100]}
grid_search4= GridSearchCV(svr2,param_grid, cv=5)
grid_search4.fit(x_train, y2_train)

print("Best Parameters : {}".format(grid_search4.best_params_))
print("Best cross_val score : {}".format(grid_search4.best_score_))

Best Parameters : {'C': 100, 'gamma': 10}
Best cross_val score : 0.9414908740283418


In [42]:
svrrbf= SVR(kernel = 'rbf', C = 100, gamma = 10)
svrrbf_cool = svrrbf.fit(x_train,y2_train)
print('R-squared score (training): {:.3f}'
     .format(svrrbf_cool.score(x_train, y2_train)))
print('R-_squared score (test): {:.3f}'
     .format(svrrbf_cool.score(x_test, y2_test)))

R-squared score (training): 1.000
R-_squared score (test): 0.947


#### Knn Regressor

In [41]:
from sklearn.neighbors import KNeighborsRegressor
knnreg = KNeighborsRegressor()
param_grid = {'n_neighbors': [3, 4, 5, 6, 7]}
grid_search_knnreg = GridSearchCV(knnreg, param_grid, cv = 5, scoring = 'r2')
grid_search_knnreg.fit(x_train, y2_train)
print("Best Parameters: {}".format(grid_search_knnreg.best_params_))
print("Best Score for KNN repressor - Cooling Load: {:.2f}".format(grid_search_knnreg.best_score_))

Best Parameters: {'n_neighbors': 7}
Best Score for KNN repressor - Cooling Load: 0.90


In [44]:
knn = KNeighborsRegressor(n_neighbors=7)
knnReg = knn.fit(x_train,y2_train)

print('R-squared score (training): {:.3f}'
     .format(knnReg.score(x_train, y2_train)))
print('R-squared score (test): {:.3f}'
     .format(knnReg.score(x_test, y2_test)))

R-squared score (training): 0.923
R-squared score (test): 0.925


#### KNN gives high and comparable train and test scores compared to all other models

#### Adaboost Gridsearch

In [42]:
from sklearn.ensemble import AdaBoostRegressor

adaboost = AdaBoostRegressor(random_state = 10)
param_grid = {'n_estimators':[10,50,100,200,500,750,1000],
              'learning_rate':[0.1, 0.05, 0.01, 0.005, 0.001, 0.0005, 0.0001]}
grid_search_adaboost= GridSearchCV(adaboost,param_grid, cv=5, scoring = 'r2')
grid_search_adaboost.fit(x_train, y2_train)

print("Best Parameters : {}".format(grid_search_adaboost.best_params_))
print("Best cross_val score : {}".format(grid_search_adaboost.best_score_))

Best Parameters : {'learning_rate': 0.1, 'n_estimators': 200}
Best cross_val score : 0.9326375047259011


#### Adaboost for Linear Regression

In [46]:
ada_lin = AdaBoostRegressor(regressor, learning_rate = 0.1, n_estimators = 200, random_state = 10)
ada_lin.fit(x_train, y2_train)

print('Adaboost for Linear Regression on training set: {:.2f}'
     .format(ada_lin.score(x_train, y2_train)))
print('Adaboost for Linear Regression on test set: {:.2f}\n'
     .format(ada_lin.score(x_test, y2_test)))

Adaboost for Linear Regression on training set: 0.88
Adaboost for Linear Regression on test set: 0.90



#### Adaboost for Linear SVM

In [47]:
ada_svrlin = AdaBoostRegressor(svrlin, learning_rate = 0.1, n_estimators = 200, random_state = 10)
ada_svrlin.fit(x_train, y2_train)

print('Adaboost for Linear SVM on training set: {:.2f}'
     .format(ada_svrlin.score(x_train, y2_train)))
print('Adaboost for Linear SVM on test set: {:.2f}\n'
     .format(ada_svrlin.score(x_test, y2_test)))

Adaboost for Linear SVM on training set: 0.88
Adaboost for Linear SVM on test set: 0.90



#### Adaboost for Kernel SVM

In [48]:
ada_svrrbf = AdaBoostRegressor(svrrbf, learning_rate = 0.1, n_estimators = 200, random_state = 10)
ada_svrrbf.fit(x_train, y2_train)

print('Adaboost for Kernel SVM on training set: {:.2f}'
     .format(ada_svrrbf.score(x_train, y2_train)))
print('Adaboost for Kernel SVM on test set: {:.2f}\n'
     .format(ada_svrrbf.score(x_test, y2_test)))

Adaboost for Kernel SVM on training set: 1.00
Adaboost for Kernel SVM on test set: 0.94



#### Adaboost for KNN

In [49]:
ada_knn = AdaBoostRegressor(knn, learning_rate = 0.1, n_estimators = 200, random_state = 10)
ada_knn.fit(x_train, y2_train)

print('Adaboost for KNN on training set: {:.2f}'
     .format(ada_knn.score(x_train, y2_train)))
print('Adaboost for KNN on test set: {:.2f}\n'
     .format(ada_knn.score(x_test, y2_test)))


Adaboost for KNN on training set: 0.95
Adaboost for KNN on test set: 0.92



#### With comparable train and test scores, KNN comes as the best model for ADAboost as well.

#### Bagging Gridsearch

In [43]:
from sklearn.ensemble import BaggingRegressor

bagging = BaggingRegressor(random_state = 10)
param_grid = {'max_samples':[10,50,100]}
grid_search_bagging = GridSearchCV(bagging, param_grid, cv=5, scoring = 'r2')
grid_search_bagging.fit(x_train, y2_train)

print("Best Parameters : {}".format(grid_search_bagging.best_params_))
print("Best cross_val score : {}".format(grid_search_bagging.best_score_))

Best Parameters : {'max_samples': 100}
Best cross_val score : 0.957887237332568


#### Bagging for Linear Regression

In [51]:
bag_lin= BaggingRegressor(regressor, max_samples = 100, random_state = 10)
bag_lin.fit(x_train, y2_train)

print('Bagging for Linear Regression on training set: {:.2f}'
     .format(bag_lin.score(x_train, y2_train)))
print('Bagging for Linear Regression on test set: {:.2f}\n'
     .format(bag_lin.score(x_test, y2_test)))

Bagging for Linear Regression on training set: 0.88
Bagging for Linear Regression on test set: 0.90



#### Bagging for Linear SVM

In [52]:
bag_svrlin= BaggingRegressor(svrlin, max_samples = 100, random_state = 10)
bag_svrlin.fit(x_train, y2_train)

print('Bagging for Linear SVM on training set: {:.2f}'
     .format(bag_svrlin.score(x_train, y2_train)))
print('Bagging for Linear SVM on test set: {:.2f}\n'
     .format(bag_svrlin.score(x_test, y2_test)))

Bagging for Linear SVM on training set: 0.86
Bagging for Linear SVM on test set: 0.90



#### Bagging for Kernel SVM

In [53]:
bag_svrrbf= BaggingRegressor(svrrbf, max_samples = 100, random_state = 10)
bag_svrrbf.fit(x_train, y2_train)

print('Bagging for Kernel SVM on training set: {:.2f}'
     .format(bag_svrrbf.score(x_train, y2_train)))
print('Bagging for Kernel SVM on test set: {:.2f}\n'
     .format(bag_svrrbf.score(x_test, y2_test)))

Bagging for Kernel SVM on training set: 0.75
Bagging for Kernel SVM on test set: 0.65



#### Bagging for KNN

In [54]:
bag_knn= BaggingRegressor(knn, max_samples = 100, random_state = 10)
bag_knn.fit(x_train, y2_train)

print('Bagging for Kernel SVM on training set: {:.2f}'
     .format(bag_knn.score(x_train, y2_train)))
print('Bagging for Kernel SVM on test set: {:.2f}\n'
     .format(bag_knn.score(x_test, y2_test)))

Bagging for Kernel SVM on training set: 0.88
Bagging for Kernel SVM on test set: 0.90



#### KNN, Linear SVM and OLS have similar train and test scores. Looking across all methods, KNN seems to be consistent and hence we can use that to predict y2

#### Deep Learning Regressor

In [92]:
# create model
model = Sequential()
model.add(Dense(13, input_dim=8, kernel_initializer='normal', activation='relu'))
model.add(Dense(6, kernel_initializer = 'normal', activation = 'relu'))
model.add(Dense(1, kernel_initializer='normal'))

# Compile model
model.compile(loss='mse', optimizer='adam' , metrics = ['mse'])

model.fit(x_train, y2_train, epochs = 100, batch_size = 20)



Epoch 1/100
691/691 [==============================] - 3s 4ms/step - loss: 696.5157 - mean_squared_error: 696.5157
Epoch 2/100
691/691 [==============================] - 0s 135us/step - loss: 690.9538 - mean_squared_error: 690.9538
Epoch 3/100
691/691 [==============================] - 0s 138us/step - loss: 673.6300 - mean_squared_error: 673.6300
Epoch 4/100
691/691 [==============================] - 0s 129us/step - loss: 628.9118 - mean_squared_error: 628.9118
Epoch 5/100
691/691 [==============================] - 0s 138us/step - loss: 543.5383 - mean_squared_error: 543.5383
Epoch 6/100
691/691 [==============================] - 0s 113us/step - loss: 417.1099 - mean_squared_error: 417.1099
Epoch 7/100
691/691 [==============================] - 0s 489us/step - loss: 271.9488 - mean_squared_error: 271.9488
Epoch 8/100
691/691 [==============================] - 0s 138us/step - loss: 152.9155 - mean_squared_error: 152.9155
Epoch 9/100
691/691 [==============================] - 0s 116us/st

691/691 [==============================] - 0s 138us/step - loss: 12.0692 - mean_squared_error: 12.0692
Epoch 72/100
691/691 [==============================] - 0s 136us/step - loss: 12.0361 - mean_squared_error: 12.0361
Epoch 73/100
691/691 [==============================] - 0s 128us/step - loss: 12.0300 - mean_squared_error: 12.0300
Epoch 74/100
691/691 [==============================] - 0s 113us/step - loss: 12.0224 - mean_squared_error: 12.0224
Epoch 75/100
691/691 [==============================] - 0s 125us/step - loss: 12.0016 - mean_squared_error: 12.0016
Epoch 76/100
691/691 [==============================] - 0s 113us/step - loss: 11.9737 - mean_squared_error: 11.9737
Epoch 77/100
691/691 [==============================] - 0s 123us/step - loss: 11.9275 - mean_squared_error: 11.9275
Epoch 78/100
691/691 [==============================] - 0s 129us/step - loss: 12.0240 - mean_squared_error: 12.0240
Epoch 79/100
691/691 [==============================] - 0s 482us/step - loss: 11.9758

In [93]:
from sklearn.metrics import r2_score

y2_train_predict = model.predict(x_train)
y2_test_predict = model.predict(x_test)

print('Train r2: {:.2f}'.format(r2_score(y2_train, y2_train_predict)))
print('Test r2: {:.2f}'.format(r2_score(y2_test, y2_test_predict)))

Train r2: 0.87
Test r2: 0.88


#### Since KNN gives a high test score with comparable training score, we can use KNN to predict y2 values.

## Define the overall load into three classes 

#### Adding heating and cooling loads

In [94]:
data1 = data
data1['Total load'] = data1['Y1'] + data1['Y2']
data2 = data1.drop(["Y1","Y2"], axis = 1)

In [45]:
data2.describe()

X1          X2          X3          X4         X5          X6  \
count  768.000000  768.000000  768.000000  768.000000  768.00000  768.000000   
mean     0.764167  671.708333  318.500000  176.604167    5.25000    3.500000   
std      0.105777   88.086116   43.626481   45.165950    1.75114    1.118763   
min      0.620000  514.500000  245.000000  110.250000    3.50000    2.000000   
25%      0.682500  606.375000  294.000000  140.875000    3.50000    2.750000   
50%      0.750000  673.750000  318.500000  183.750000    5.25000    3.500000   
75%      0.830000  741.125000  343.000000  220.500000    7.00000    4.250000   
max      0.980000  808.500000  416.500000  220.500000    7.00000    5.000000   

               X7         X8  Total load  
count  768.000000  768.00000  768.000000  
mean     0.234375    2.81250   46.894961  
std      0.133221    1.55096   19.484947  
min      0.000000    0.00000   16.950000  
25%      0.100000    1.75000   28.750000  
50%      0.250000    3.00000   40.970000  
75%      0.400000    4.00000   64.335000  
max      0.400000    5.00000   89.950000

In [96]:
#First initialize the categorical binary matrix as 0s
data2['Class1']=0
data2['Class2']=0
data2['Class3']=0

#Then, activate the classification based on Y3 values
data2.loc[(data2['Total load'] < 28.75),'Class3'] = 1 
data2.loc[((data2['Total load'] >= 28.75) & (data2['Total load'] <= 64.335)),'Class2'] = 1 
data2.loc[(data2['Total load'] > 64.335),'Class1'] = 1

In [97]:
data2.describe()

X1          X2          X3          X4         X5          X6  \
count  768.000000  768.000000  768.000000  768.000000  768.00000  768.000000   
mean     0.764167  671.708333  318.500000  176.604167    5.25000    3.500000   
std      0.105777   88.086116   43.626481   45.165950    1.75114    1.118763   
min      0.620000  514.500000  245.000000  110.250000    3.50000    2.000000   
25%      0.682500  606.375000  294.000000  140.875000    3.50000    2.750000   
50%      0.750000  673.750000  318.500000  183.750000    5.25000    3.500000   
75%      0.830000  741.125000  343.000000  220.500000    7.00000    4.250000   
max      0.980000  808.500000  416.500000  220.500000    7.00000    5.000000   

               X7         X8  Total load      Class1      Class2      Class3  
count  768.000000  768.00000  768.000000  768.000000  768.000000  768.000000  
mean     0.234375    2.81250   46.894961    0.250000    0.500000    0.250000  
std      0.133221    1.55096   19.484947    0.433295    0.500326    0.433295  
min      0.000000    0.00000   16.950000    0.000000    0.000000    0.000000  
25%      0.100000    1.75000   28.750000    0.000000    0.000000    0.000000  
50%      0.250000    3.00000   40.970000    0.000000    0.500000    0.000000  
75%      0.400000    4.00000   64.335000    0.250000    1.000000    0.250000  
max      0.400000    5.00000   89.950000    1.000000    1.000000    1.000000

## Predicting Classes using Deep Learning

In [98]:
# fix random seed for reproducibility
np.random.seed(10)

In [99]:
x = data2[['X1','X2','X3','X4','X5','X6','X7','X8']]
y = data2[['Class1','Class2','Class3']]

x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.1,random_state=10)
x.shape

(768, 8)

In [100]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.fit_transform(x_test)

In [101]:
# create model
model = Sequential()
model.add(Dense(12, input_dim=8, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(3, activation='sigmoid'))

In [106]:
# Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [107]:
# Fit the model
model.fit(x_train, y_train, epochs=100, batch_size=20)

Epoch 1/100
691/691 [==============================] - 3s 5ms/step - loss: 0.9669 - acc: 0.7115
Epoch 2/100
691/691 [==============================] - 0s 145us/step - loss: 0.5813 - acc: 0.7501
Epoch 3/100
691/691 [==============================] - 0s 143us/step - loss: 0.3845 - acc: 0.8061
Epoch 4/100
691/691 [==============================] - 0s 147us/step - loss: 0.3050 - acc: 0.8370
Epoch 5/100
691/691 [==============================] - 0s 138us/step - loss: 0.2748 - acc: 0.8548
Epoch 6/100
691/691 [==============================] - 0s 138us/step - loss: 0.2633 - acc: 0.8712
Epoch 7/100
691/691 [==============================] - 0s 147us/step - loss: 0.2562 - acc: 0.8722
Epoch 8/100
691/691 [==============================] - 0s 136us/step - loss: 0.2525 - acc: 0.8881
Epoch 9/100
691/691 [==============================] - 0s 329us/step - loss: 0.2490 - acc: 0.9026
Epoch 10/100
691/691 [==============================] - 0s 313us/step - loss: 0.2462 - acc: 0.9045
Epoch 11/100
691/691 

691/691 [==============================] - 0s 280us/step - loss: 0.1812 - acc: 0.9141
Epoch 83/100
691/691 [==============================] - ETA: 0s - loss: 0.1910 - acc: 0.918 - 0s 295us/step - loss: 0.1804 - acc: 0.9204
Epoch 84/100
691/691 [==============================] - 0s 197us/step - loss: 0.1800 - acc: 0.9165
Epoch 85/100
691/691 [==============================] - 0s 148us/step - loss: 0.1793 - acc: 0.9175
Epoch 86/100
691/691 [==============================] - 0s 126us/step - loss: 0.1796 - acc: 0.9156
Epoch 87/100
691/691 [==============================] - 0s 128us/step - loss: 0.1790 - acc: 0.9233
Epoch 88/100
691/691 [==============================] - 0s 129us/step - loss: 0.1783 - acc: 0.9209
Epoch 89/100
691/691 [==============================] - 0s 178us/step - loss: 0.1788 - acc: 0.9141
Epoch 90/100
691/691 [==============================] - 0s 170us/step - loss: 0.1773 - acc: 0.9165
Epoch 91/100
691/691 [==============================] - 0s 167us/step - loss: 0.1777

In [110]:
# evaluate the model
scores = model.evaluate(x_test, y_test)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

77/77 [==============================] - 0s 130us/step

acc: 90.48%


In [109]:
scores1 = model.evaluate(x_train, y_train)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores1[1]*100))

691/691 [==============================] - 0s 59us/step

acc: 91.80%


### The model gives a training accuracy of 91.8% and a test accuracy of 90.48% which indicates that the model will do a good job at predicting the classes for new values of heating and cooling load.

### Conclusion
#### Heating Load prediction - Base ML methods - OLS and Linear SVR gave very good results and turned out to be good models. Neural networks gave a very similar score and hence we can choose any one amongst them. Different values of batch size and epochs may give us a better result.
#### Heating Load prediction - Base ML methods - KNN gave very good results compared to other models. Neural networks gave a slightly lower training and test score. Different values of batch size and epochs may give us a better result.
#### Combined Load Multi-Class prediction: Almost equal training and test scores. Scores can be increased by tweaking values of neuron layers and other parameters as mentioned above.